MathLib/ODE/CVodeSolver.cpp

```diff
@@ -0,0 +1,353 @@
+}
+}
+
+namespace MathLib
+{
+/**
+ * This class provides concrete access to Sundials' CVode solver.
+ *
+ * OdeSolver (implicitly bounds checked, agnostic to concrete ODE solver)
+ *  |
+ *  | Dynamic polymorphism
+ *  v
+ * ConcreteOdeSolver (implicitly bounds checked, interfaces with a specific
+ * library)
+ *  |
+ *  | Forward calls, disable bounds checking, no need for templates anymore
+ *  v
+ * Implementation = CVodeSolver (no templates)
+ *  |
+ *  | Pimpl (hide implementation, do not include 3rd party libs in header)
+ *  v
+ * CVodeSolverImpl
+ *
+ * This scheme might be a general way for accessing 3rd party libraries.
+ */
+class CVodeSolverImpl
+{
+	static_assert(std::is_same<realtype, double>::value,
+	              "cvode's realtype is not the same as double");
+
+public:
+	CVodeSolverImpl(BaseLib::ConfigTree const& config,
+	                unsigned const num_equations);
+
+	friend class CVodeSolver;
+	~CVodeSolverImpl();
+
+private:
+	void setFunction(std::unique_ptr<FunctionHandles>&& f);
+
+	void preSolve();
+	void solve(const double t_end);
+
+	double const* getSolution() const { return NV_DATA_S(_y); }
+	double getTime() const { return _t; }
+	void getYDot(const double t, double const* const y, double* const y_dot);
+	void setTolerance(const double* abstol, const double reltol);
+	void setTolerance(const double abstol, const double reltol);
+	void setIC(const double t0, double const* const y0);
+
+private:
+	N_Vector _y = nullptr;
+
+	realtype _t;
+
+	N_Vector _abstol = nullptr;
+	realtype _reltol;
+
+	unsigned _num_equations;
+	void* _cvode_mem;
+
+	std::unique_ptr<FunctionHandles> _f;
+
+	int _linear_multistep_method = CV_ADAMS;
+	int _nonlinear_solver_iteration = CV_FUNCTIONAL;
+};
+
+CVodeSolverImpl::CVodeSolverImpl(const BaseLib::ConfigTree& config,
+                                 const unsigned num_equations)
+{
+	if (auto const param =
+	        config.getConfParamOptional<std::string>("linear_multistep_method"))
+	{
+		DBUG("setting linear multistep method (config: %s)", param->c_str());
+
+		if (*param == "Adams")
+		{
+			_linear_multistep_method = CV_ADAMS;
+		}
+		else if (*param == "BDF")
+		{
+			_linear_multistep_method = CV_BDF;
+		}
+		else
+		{
+			ERR("unknown linear multistep method: %s", param->c_str());
+			std::abort();
+		}
+	}
+
+	if (auto const param = config.getConfParamOptional<std::string>(
+	        "nonlinear_solver_iteration"))
+	{
+		DBUG("setting nonlinear solver iteration (config: %s)", param->c_str());
+
+		if (*param == "Functional")
+		{
+			_nonlinear_solver_iteration = CV_FUNCTIONAL;
+		}
+		else if (*param == "Newton")
+		{
+			_nonlinear_solver_iteration = CV_NEWTON;
+		}
+		else
+		{
+			ERR("unknown nonlinear solver iteration: %s", param->c_str());
+			std::abort();
+		}
+	}
+
+	_y = N_VNew_Serial(num_equations);
+	_abstol = N_VNew_Serial(num_equations);
+	_num_equations = num_equations;
+
+	_cvode_mem =
+	    CVodeCreate(_linear_multistep_method, _nonlinear_solver_iteration);
```

Comment:
cannot be checked, because a pointer is returned. However, I added an error message.